devito/types/separable.py

```diff
@@ -0,0 +1,192 @@
+
+from devito.finite_differences import Differentiable
+from devito.finite_differences.differentiable import Mul, Add
+from devito.tools import as_tuple
+from devito.types.basic import AbstractCachedSymbol
+from devito.types.dense import Function, TimeFunction
+
+__all__ = ['SeparableFunction', 'SeparableTimeFunction']
+
+
+class SeparableFunction(Differentiable, AbstractCachedSymbol):
+    """
+    A Function separable in its Dimensions, ie f(x,y,z) = f(x)*f(y)*f(z)
+
+    Take the same parameters as a Function plus
+
+    Parameters
+    ----------
+    sep: str
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated: Dimension or tuple of Dimensions
+        Dimension in which the Function is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> SeparableFunction(name='a', grid=grid)
+    a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableFunction(name='b', grid=grid, sep='sum')
+    b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableFunction(name='c', grid=grid, sep='sum', separated=x)
+    c_x(x) + c_yz(y, z)
+    """
+    def __init__(self, *args, **kwargs):
+        # Is it sum or product separable
+        self._sep_op = spearation_op[kwargs.get('sep', 'prod')]
+
+        # Conventional Function inputs
+        self._grid = kwargs.get('grid')
+        self._name = kwargs.get('name')
+        self._space_order = kwargs.get('space_order', 1)
+
+        # Initialize subfunctions
+        # Separable dimensions
+        sep = self.separated_dimensions(**kwargs)
+        nonseparated = tuple(d for d in self.grid.dimensions if d not in sep)
+        dims = tuple(d for d in sep) + (nonseparated,)
+        self._dims = dims
+
+        func_list = [self.function(d, **kwargs) for d in sep]
+        if len(nonseparated) > 0:
+            func_list.append(self.function(nonseparated, **kwargs))
+
+        expr = self._sep_op(*func_list)
+
+        dims = tuple(d for d in sep) + (nonseparated,)
+        self._subfunctions = {d: f for d, f in zip(dims, func_list)}
+        self._dims = dims
+        self._expr = expr
+        self._fd = self.expr._fd
+
+    def function(self, dimensions, **kwargs):
+        dims = as_tuple(dimensions)
+        func_type = Function
+        names = "".join(d.name for d in dims)
+        shape = tuple(self.grid.dimension_map[d].loc for d in dims)
```

Comment:
so, currently, SeparableFunctions won't work if there's a generic dimensions that doesn't appear in ``grid`` , correct?

Reply:
Yes good catch, need to add that case